documentation/OpenMP-semantics.md

```diff
@@ -0,0 +1,662 @@
+# OpenMP Semantic Analysis
+
+## OpenMP for F18
+
+1. Define and document the parse tree representation for
+    * Directives (listed below)
+    * Clauses (listed below)
+    * Documentation
+1. All the directives and clauses need source provenance for messages
+1. Define and document how an OpenMP directive in the parse tree
+will be represented as the parent of the statement(s)
+to which the directive applies.
+The parser itself will not be able to construct this representation;
+there will be subsequent passes that do so
+just like for example _do-stmt_ and _do-construct_.
+1. Define and document the symbol table extensions
+1. Define and document the module file extensions
+
+
+### Directives
+
+OpenMP divides directives into three categories as follows.
+The directives that are in the same categories share some characteristics.
+
+
+
+#### Declarative directives
+
+An OpenMP directive may only be placed in a declarative context.
+A declarative directive results in one or more declarations only;
+it is not associated with the immediate execution of any user code.
+
+List of existing ones:
+* declare simd
+* declare target
+* threadprivate
+* declare reduction
+
+There is a parser node for each of these directives and
+the parser node saves information associated with the directive,
+for example,
+the name of the procedure-name in the `declare simd` directive.
+
+Each parse tree node keeps source provenance,
+one for the directive name itself and
+one for the entire directive starting from the directive name.
+
+A top-level class, `OpenMPDeclarativeConstruct`,
+holds all four of the node types as discriminated unions
+along with the source provenance for the entire directive
+starting from `!$OMP`.
+
+In `parser-tree.h`,
+`OpenMPDeclarativeConstruct` is part
+of the `SpecificationConstruct` and `SpecificationPart`
+in F18 because
+a declarative directive can only be placed in the specification part
+of a Fortran program.
+
+All the `Names` or `Designators` associated
+with the declarative directive will be resolved in later phases.
+
+#### Executable directives
+
+An OpenMP directive that is **not** declarative.
+That is, it may only be placed in an executable context.
+It contains stand-alone directives and constructs
+that are associated with code blocks.
+The stand-alone directive is described in the next section.
+
+The constructs associated with code blocks listed below
+share a similar structure:
+_Begin Directive_, _Clause List_, _Code Block_, _End Directive_.
+The _End Directive_ is optional for constructs
+like Loop-associated constructs.
+
+* Block-associated constructs (`OpenMPBlockConstruct`)
+* Loop-associated constructs (`OpenMPLoopConstruct`)
```

Comment:
Is there any special handling for OpenMP loop constructs like collapse(n) which affects not just the immediate loop but the outermost n loops?

Reply:
Would you mind being more specific on the "special handling"? If you are talking about `collapse(n)` on the parse tree level, then no. Parse tree node is not designed for this. For `collapse(n)` or `ordered(n)`, during Semantics, we will do the `associated loop(s)` check based on the parameter `n` and the information of `associated loop(s)` (# of nesting levels, etc.) will be stored in the context of `OpenMPLoopConstruct`. More details are still TBD, for example, OpenMP 5.0 does not require perfectly nested loops any more if `collapse(n)` is present.

Reply:
Thanks for the reply.
I was asking because with constructs like collapse there is a relation between the OpenMP construct and n nested loops (not just the next loop). 
When this loop will be lowered to MLIR we might want to tag those n loops with some information so that an optimization pass can collapse them.

Reply:
The nested loops will be children of the directive/clause so the lowering code ought to be able to apply the collapse clause.  Makes sense that the collapse could be written as an MLIR pass.

Reply:
The OpenMP `collapse` clause ought to be associated with the OpenMP construct in the parse tree.  The lowering of that to the OpenMP dialect of MLIR needs some design, obviously.  One possible approach could be defining a collapse attribute to some OpenMP operation that applies to its Region, which in turn contains the loop nest.